Tests/AuthenticationInterceptorTests.swift

```diff
@@ -0,0 +1,663 @@
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+@testable import Alamofire
+import Foundation
+import XCTest
+
+class AuthenticationInterceptorTestCase: BaseTestCase {
+
+    // MARK: - Helper Types
+
+    struct OAuthCredential: AuthenticationCredential {
+        let accessToken: String
+        let refreshToken: String
+        let userID: String
+        let expiration: Date
+
+        let requiresRefresh: Bool
+
+        init(
+            accessToken: String = "a0",
+            refreshToken: String = "r0",
+            userID: String = "u0",
+            expiration: Date = Date(),
+            requiresRefresh: Bool = false)
+        {
+            self.accessToken = accessToken
+            self.refreshToken = refreshToken
+            self.userID = userID
+            self.expiration = expiration
+            self.requiresRefresh = requiresRefresh
+        }
+    }
+
+    enum OAuthError: Error {
+        case refreshNetworkFailure
+    }
+
+    class OAuthAuthenticator: Authenticator {
+        private(set) var applyCount = 0
+        private(set) var refreshCount = 0
+        private(set) var didRequestFailDueToAuthErrorCount = 0
+        private(set) var isRequestAuthenticatedWithCredentialCount = 0
+
+        let refreshResult: Result<OAuthCredential, Error>?
+        let lock = NSLock()
+
+        init(refreshResult: Result<OAuthCredential, Error>? = nil) {
+            self.refreshResult = refreshResult
+        }
+
+        func apply(_ credential: OAuthCredential, to urlRequest: inout URLRequest) {
+            lock.lock() ; defer { lock.unlock() }
+
+            applyCount += 1
+
+            let authorization = HTTPHeader.authorization(bearerToken: credential.accessToken)
+            urlRequest.addValue(authorization.value, forHTTPHeaderField: authorization.name)
+        }
+
+        func refresh(_ credential: OAuthCredential, completion: @escaping (Result<OAuthCredential, Error>) -> Void) {
+            lock.lock() ; defer { lock.unlock() }
+
+            refreshCount += 1
+
+            let refreshResult = self.refreshResult ?? .success(
+                OAuthCredential(
+                    accessToken: "a\(refreshCount)",
+                    refreshToken: "a\(refreshCount)",
+                    userID: "u1",
+                    expiration: Date()
+                )
+            )
+
+            DispatchQueue.global(qos: .utility).asyncAfter(deadline: .now() + 0.1) { completion(refreshResult) }
+        }
+
+        func didRequest(_ urlRequest: URLRequest, with response: HTTPURLResponse, failDueToAuthenticationError error: Error) -> Bool {
+            lock.lock() ; defer { lock.unlock() }
+
+            didRequestFailDueToAuthErrorCount += 1
+
+            return response.statusCode == 401
```

Comment:
Then entire test suite considers a 401 status code to be the trigger for refresh.